diff -Nur -Nur wine-d3d9-4.21.orig/programs/winecfg/resource.h wine-d3d9-4.21/programs/winecfg/resource.h
--- wine-d3d9-4.21.orig/programs/winecfg/resource.h	2019-12-02 13:42:13.169900792 +0800
+++ wine-d3d9-4.21/programs/winecfg/resource.h	2019-12-02 14:12:54.669874489 +0800
@@ -219,6 +219,7 @@
 #define IDC_ENABLE_EAX                  9003
 #define IDC_ENABLE_HIDEWINE             9004
 #define IDC_ENABLE_GTK3                 9005
+#define IDC_ENABLE_NATIVE_D3D9          9006
 
 /* About tab */
 #define IDC_ABT_OWNER                8432
diff -Nur -Nur wine-d3d9-4.21.orig/programs/winecfg/staging.c wine-d3d9-4.21/programs/winecfg/staging.c
--- wine-d3d9-4.21.orig/programs/winecfg/staging.c	2019-12-02 13:42:13.169900792 +0800
+++ wine-d3d9-4.21/programs/winecfg/staging.c	2019-12-02 14:12:54.673874489 +0800
@@ -121,6 +121,23 @@
 #endif
 }
 
+/*
+ * Gallium nine
+ */
+static BOOL nine_get(void)
+{
+    BOOL ret;
+    char *value = get_reg_key(config_key, keypath("DllRedirects"), "d3d9", NULL);
+    ret = (value && !strcmp(value, "d3d9-nine.dll"));
+    HeapFree(GetProcessHeap(), 0, value);
+    return ret;
+}
+
+static void nine_set(BOOL status)
+{
+    set_reg_key(config_key, keypath("DllRedirects"), "d3d9", status ? "d3d9-nine.dll" : NULL);
+}
+
 static void load_staging_settings(HWND dialog)
 {
     CheckDlgButton(dialog, IDC_ENABLE_CSMT, csmt_get() ? BST_CHECKED : BST_UNCHECKED);
@@ -128,6 +145,7 @@
     CheckDlgButton(dialog, IDC_ENABLE_EAX, eax_get() ? BST_CHECKED : BST_UNCHECKED);
     CheckDlgButton(dialog, IDC_ENABLE_HIDEWINE, hidewine_get() ? BST_CHECKED : BST_UNCHECKED);
     CheckDlgButton(dialog, IDC_ENABLE_GTK3, gtk3_get() ? BST_CHECKED : BST_UNCHECKED);
+    CheckDlgButton(dialog, IDC_ENABLE_NATIVE_D3D9, nine_get() ? BST_CHECKED : BST_UNCHECKED);
 
 #ifndef HAVE_VAAPI
     disable(IDC_ENABLE_VAAPI);
@@ -135,6 +153,9 @@
 #if !defined(HAVE_GTK_GTK_H) || !defined(SONAME_LIBGTK_3)
     disable(IDC_ENABLE_GTK3);
 #endif
+#if !defined(HAVE_D3D9NINE)
+    disable(IDC_ENABLE_NATIVE_D3D9);
+#endif
 }
 
 INT_PTR CALLBACK StagingDlgProc(HWND hDlg, UINT uMsg, WPARAM wParam, LPARAM lParam)
@@ -180,6 +201,10 @@
             gtk3_set(IsDlgButtonChecked(hDlg, IDC_ENABLE_GTK3) == BST_CHECKED);
             SendMessageW(GetParent(hDlg), PSM_CHANGED, 0, 0);
             return TRUE;
+        case IDC_ENABLE_NATIVE_D3D9:
+            nine_set(IsDlgButtonChecked(hDlg, IDC_ENABLE_NATIVE_D3D9) == BST_CHECKED);
+            SendMessageW(GetParent(hDlg), PSM_CHANGED, 0, 0);
+            return TRUE;
         }
         break;
     }
diff -Nur -Nur wine-d3d9-4.21.orig/programs/winecfg/winecfg.rc wine-d3d9-4.21/programs/winecfg/winecfg.rc
--- wine-d3d9-4.21.orig/programs/winecfg/winecfg.rc	2019-12-02 13:42:13.169900792 +0800
+++ wine-d3d9-4.21/programs/winecfg/winecfg.rc	2019-12-02 14:12:54.677874489 +0800
@@ -324,6 +324,7 @@
     CONTROL     "Enable Environmental Audio E&xtensions (EAX)",IDC_ENABLE_EAX,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,16,70,230,8
     CONTROL     "&Hide Wine version from applications",IDC_ENABLE_HIDEWINE,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,16,85,230,8
     CONTROL     "Enable &GTK3 Theming",IDC_ENABLE_GTK3,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,16,100,230,8
+    CONTROL     "Enable Gallium &Nine for better D3D9 graphic performance",IDC_ENABLE_NATIVE_D3D9,"Button",BS_AUTOCHECKBOX | WS_TABSTOP,16,115,230,8
 END
 
 LANGUAGE LANG_NEUTRAL, SUBLANG_NEUTRAL
